ramp4-pcar4/ramp4-pcar4
# Implement projection change + extent sets

## Closes #144, Closes #655

## Changes in this PR
- [FEAT] Implemented projection change 
	- Note that this implementation does not require a full map reset and will maintain the state of the map
	- When the main map changes projection, the Overview map will adapt its basemap to match the projection
- [FEAT] Implemented extent sets
	- Extent sets have replaced the old `extent` property in the config 
	- Implemented in `extent-set.ts` under `src > geo > api > map`
	- Also updated all starter scripts to use new extent and lod set config schemas 
- [FEAT] Architecture changes to the map APIs
	- `CommonMapAPI` now supports the following methods that are used/overridden by `MapAPI` and `OverviewMapAPI`:
		- `createMap`: Creates an ESRI map and map view given a `RampMapConfig` and `targetDiv` element.
		- `createMapView [abstract]`: Creates an ESRI map view. Can provide a basemap to be used when creating the view.
		- `refreshMap`: Deletes the ESRI map view and recreates it. Can provide a basemap as well.
		- `findBasemap`: Fetches the basemap from the api's local basemap list or config.
		- `setBasemap [abstract]`: Sets the basemap. Will refresh the map view if projection change is required.
		- `applyBasemap`: Applies the given basemap to the ESRI map.
		- Other utility methods that are inherited like `geomToMapSR`,  `zoomMapTo`, `getZoomLevel`, etc.
	
## [Demo](http://ramp4-app.azureedge.net/demo/users/sharvenp/projection-change/host/index.html)
### Steps to test

1. Open the basemap panel and switch to different basemaps - the main map and overview map should update
2. Test layer functions (add, reload, remove, identify)
3. Test geosearch
4. Test export 
5. Test layer reorder
6. Test the home button
7. Check if the north pole appears in lambert basemaps
8. More test cases...?


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/850)
